src/adapter-nextjs/processRequest.ts

```diff
@@ -1,31 +1,52 @@
 import type { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
 
-import { getRedirectToCanonicalLocale } from '../intl';
-import { BasePageProps } from '../types';
+import { getShortestLocaleCode } from '../data-fetching';
+import { getRedirectToCanonicalLocale, LocaleObject } from '../intl';
+import { PageProps, ServerSidePageProps } from '../types';
 
-export function processRequest<Props extends BasePageProps>(
+// TODO: Try to make the typings even more robust
+function extractServerSideProps<Props>(props: Props & ServerSidePageProps): {
+    pageProps: Props;
+    serverSideProps: ServerSidePageProps;
+} {
+    const { localeResolved, ...pageProps } = props;
+
+    return {
+        pageProps: pageProps as unknown as Props,
+        serverSideProps: { localeResolved },
+    };
+}
+
+export function processRequest<Props>(
     context: GetServerSidePropsContext,
-    basePageProps: BasePageProps,
+    props: Props & PageProps & ServerSidePageProps,
     canonicalUrl?: string,
-    customProps?: Omit<Props, keyof BasePageProps>,
 ): GetServerSidePropsResult<Props> {
-    const { locale, query } = context;
+    const { locale: nextLocale, query } = context;
+
+    const { pageProps, serverSideProps } = extractServerSideProps(props);
```

Comment:
What's the point of this function? Can't you just do
```typescript
const { localeResolved, ...pageProps } = props;

if (!localeResolved) {
    return { notFound: true };
}

if (canonicalUrl) {
    const { languages, localeCode } = pageProps.newsroomContextProps;
    const currentLocale = LocaleObject.fromAnyCode(localeCode);
    const shortestLocaleCode = getShortestLocaleCode(languages, currentLocale);

    const redirect = getRedirectToCanonicalLocale(
        shortestLocaleCode,
        nextLocale,
        canonicalUrl,
        query,
    );
    if (redirect) {
        return { redirect };
    }
}

return {
    props: pageProps as Props,
};
```

This way `localeResolved` will be dropped from the `pageProps` and then you just return them. Typescript doesn't complain so it should work?

Reply:
This is a bit of premature refactor, initially it was the same way as in your suggestion, but the point is, if we change those types in the future, there will be no warnings or error from TS, which is not optimal. The function helps to check at least the `ServerSidePageProps` type.
But maybe I'm overthinking this :D

Reply:
Well the worst thing that could happen would be that some of the server-side props would be passed through to the component, right?

So I'd say it sounds like overthinking.. 😄

Reply:
You're right, although this would be hard to spot and debug if you don't know where to look at. But let's leave it for future us :D

Reply:
Oh, the unknown part actually fixed itself somehow. I guess it was due to switch from `extends` interface to an union type. So it's all good now!